### PR TITLE
improve correlation for EventSource events in SslStream

### PR DESCRIPTION
**Pull request by wfurt**
I bump to this while working on #36192. The test would fail only when running whole set and never when running alone. I was trying to use tracing too get clue what is going on but it is pretty difficult to filter out relevant events even when I have package capture.

To get a hint I would get something like:
```

<Event MSec=  "6322.6179" PID="20040" PName="Process(20040)" TID="20035" EventName="Enter" ProviderName="Microsoft-System-Net-Security" thisOrContextObject="SecureChannel#15015783" memberName=".ctor" parameters="(testservereku.contoso.com, X509CertificateColle
ction(1))"/>
<Event MSec=  "6322.6445" PID="20040" PName="Process(20040)" TID="20035" EventName="SecureChannelCtor" ProviderName="Microsoft-System-Net-Security" hostname="testservereku.contoso.com" secureChannelHash="15,015,783" clientCertificatesCount="1" encryptionPolicy="0”/>
<Event MSec=  "6322.6704" PID="20040" PName="Process(20040)" TID="20035" EventName="Exit" ProviderName="Microsoft-System-Net-Security" thisOrContextObject="SecureChannel#15015783" memberName=".ctor" result="”/>
```

This is pretty difficult to use because:
-  SecureChannel is internal and it is hard for caller to predict the hash
- the hash have two flavors and that makes text search difficult
- the hostname may be clue but that is not sufficient when there are many streams to the same destination. 
- There is no good way how to correlate that to external sources.

To improve that, I added trace to SslStream constructor itself with also some info about provided inner stream. 
```
<Event MSec=  "3968.7405" PID="5648" PName="Process(5648)" TID="5643" EventName="SslStreamCtor" ProviderName="Microsoft-System-Net-Security" thisOrContextObject="SslStream#55177178" localId="VirtualNetworkStream#37761023" remoteId=""/>
<Event MSec=  "3962.6993" PID="5648" PName="Process(5648)" TID="5642" EventName="SslStreamCtor" ProviderName="Microsoft-System-Net-Security" thisOrContextObject="SslStream#8259783" localId="127.0.0.1:42365" remoteId="127.0.0.1:39410"/>
```

and I added sslStream to SecureChannel constructor so it is possible to link them together.

```
<Event MSec=  "3970.3398" PID="5648" PName="Process(5648)" TID="5663" EventName="SecureChannelCtor" ProviderName="Microsoft-System-Net-Security" sslStream="SecureChannel#9578510" hostname="testservereku.contoso.com" secureChannelHash="9,578,510" clientCertificatesCount="1" encryptionPolicy="0"/>
```

and I added Enter/Exit entry to CloseInternal() so we can track disposal. This could possibly do to Dispose() function as well or be converted to  NetEventSource.Log to match better the constructor event. 

contributes to #2359

### CODE REVIEW

File: src/libraries/System.Net.Security/src/System/Net/Security/NetEventSource.Security.cs [hunk 1]
```diff
@@ -378,5 +414,50 @@ private unsafe void WriteEvent(int eventId, string arg1, int arg2, int arg3, int
                 }
             }
         }
+
+        [NonEvent]
+        private unsafe void WriteEvent(int eventId, string arg1, string arg2, int arg3, int arg4, int arg5)
```
Comment (davidsh) [inline src/libraries/System.Net.Security/src/System/Net/Security/NetEventSource.Security.cs:419]: We usually put these helpers into the base NetEventSource class.

See examples: https://github.com/dotnet/runtime/blob/master/src/libraries/Common/src/System/Net/Logging/NetEventSource.Common.cs

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs [hunk 2]
```diff
@@ -130,6 +130,9 @@ private void SetException(Exception e)
         //
         private void CloseInternal()
         {
+            if (NetEventSource.IsEnabled)
```
Comment (davidsh) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs:133]: nit: In most of our sources, these simple trace statements are on a single line:

```c#
if (NetEventSource.IsEnabled) NetEventSource.Enter(this);
```

File: src/libraries/System.Net.Security/src/System/Net/Security/NetEventSource.Security.cs (hunk 1 above)
Comment (wfurt) [reply src/libraries/System.Net.Security/src/System/Net/Security/NetEventSource.Security.cs:419]: Should I move the other WriteEvent() as well? I saw them in both places...

Comment (davidsh) [reply src/libraries/System.Net.Security/src/System/Net/Security/NetEventSource.Security.cs:419]: >I saw them in both places...

What are "both" places?

I think all the `WriteEvent` helpers should be in the base class.

Comment (wfurt) [reply src/libraries/System.Net.Security/src/System/Net/Security/NetEventSource.Security.cs:419]: https://github.com/dotnet/runtime/blob/184f49b20ee9fc6c2e6649e22750cb6c54ffdc53/src/libraries/System.Net.Security/src/System/Net/Security/NetEventSource.Security.cs#L324-L325

There is already WriteEvent() overload in NetEventSource.Security.cs in addition to NetEventSource.Common.cs

Having everything in base class make sense to me.

Comment (wfurt) [reply src/libraries/System.Net.Security/src/System/Net/Security/NetEventSource.Security.cs]: updated.

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs (hunk 2 above)
Comment (Wraith2) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: That's contrary to the coding guidelines for the repo. I disliked the unbraced style so I checked and unbraced multi line if is permitted but single line is not. See https://github.com/dotnet/runtime/blob/master/docs/coding-guidelines/coding-style.md item 17.

**Comment by wfurt** [comment]
could you please take another look @davidsh?

**Comment by davidsh** [comment]
>could you please take another look @davidsh?

Sure. I will do it tomorrow.

File: src/libraries/System.Net.Security/src/System/Net/Security/NetEventSource.Security.cs [hunk 3]
```diff
@@ -54,16 +57,49 @@ public void SspiPackageNotFound(string packageName)
...
+
+                if (localId == null)
+                {
+                    localId = IdOf(innerStream);
+                }
+
+                SslStreamCtor(IdOf(sslStream), localId, remoteId);
             }
         }
+        [Event(SslStreamCtorId, Keywords = Keywords.Default, Level = EventLevel.Informational)]
```
Comment (davidsh) [inline src/libraries/System.Net.Security/src/System/Net/Security/NetEventSource.Security.cs:87]: nit: add blank line above this.

**Review by davidsh — APPROVED** [review]
LGTM

**Pull request merged**